docs/proposals/20210705-bastion-vms.md

```diff
@@ -0,0 +1,198 @@
+
+Those solutions might not cover all use cases. As a matter of example:
+
+- Using one master node as a jump host might be not compliant to the security policies for some companies.
+- Using `AzureBastion` might be inpractical or unusable in some use cases (machine to machine access would be nearly impossible).
+
+### Goals
+
+- Allow users to deploy virtual machines as bastion hosts in multiple availability zones using a custom cloud init file.
+
+### Non-Goals/Future Work
+
+- Make it possible to disable the default behaviour of exposing one master node to the internet.
+
+## Proposal
+
+To adhere to the KISS principle the idea is to reuse existing reconciliation controllers for creating the `Bastion VM(s)`.
+In practice that means creating 3 resources:
+
+1. A `Secret` holding the `cloud-init` config file to bootstrap the bastion VM with the needed config (has to be provided by user and can't be defaulted in any way)
+2. An `AzureMachineTemplate` for each bastion that describes how the VM should be (size, OS, disks, etc). Can be defaulted.
+3. A `MachineDeployment` to bind the two aformentioned resources and make CAPI/CAPZ controllers reconcile and create the VMs. Can be defaulted.
+
+The suggestion is to add a new subfield to the `AzureCluster/spec/azureBastion` field like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: AzureCluster
+metadata:
+  name: cluster-name
+  namespace: default
+spec:
+  azureBastion:
+    bastionHosts:
+    - bootstrapDataSecretName: "<secret name>"
+      azureMachineTemplate: {}
+      replicas: 2
+```
+
+With the above information, the `AzureCluster controller` can create an `AzureMachineTemplate` and a `MachineDeployment` and then the creation of VMs should happen automagically.
+
+Why a slice of `bastionHosts` rather than a single value? Because having multiple bastion hosts only makes sense if they are placed in different availability zones.
+Since the `failureDomain` field is a scalar in the `AzureMachineTemplate` CR, it's needed to have different templates for different availability zones.
+
+![Figure 1](./images/bastion-vms/fig1.png "Architecture of the proposal")
+
+### User Stories
+
+#### Story 1
+
+Frank is a Platform Engineer for a company offering Kubernetes as a Service.
+The company has a lot of kubernetes clusters and Frank, together with his colleagues, needs to access the machines
+forming the Kubernetes cluster via SSH in order to provide support to his company's customers.
+
+For security policy compliance reasons, the master node(s) of the clusters can't expose the SSH port to the public internet.
+
+Frank writes a `MutatingWebhook` that ensures every `AzureCluster` CR contains a `spec.bastionSpec.bastionHosts` to set up 
+one bastion host for every cluster.
+
+This way he ensures all Clusters have the same setup in terms of Bastions and he can create a tool to easily SSH into
```

Comment:
Are all of the SSH keys in the cloud-init?

Reply:
Potentially, but this is user choice.
For example, we (Giant Swarm) would not use ssh keys but use Hashicorp's Vault for SSH purposes.
The whole point of asking the client to give a full cloud init secret is to have max flexibility for virtually any use case

Reply:
I see you point about having flexibility.

How do you envision the behavior when an operator will mutate or replace the cloud-init secret?

Reply:
The bastion hosts should be rolled in my opinion.
In general, I would stick to what CAPZ is doing with node pool nodes in this case.